// famous/dom-renderers/events/ScrollEvent.js
'use strict';

var Event = require('./Event');

/**
 * See [UI Events (formerly DOM Level 3 Events)](http://www.w3.org/TR/2015/WD-uievents-20150428).
 *
 * @class UIEvent
 * @augments Event
 *
 * @param  {Event} ev   The native DOM event.
 */
function ScrollEvent(ev) {
    // [Constructor(DOMString type, optional UIEventInit eventInitDict)]
    // interface UIEvent : Event {
    //     readonly    attribute Window? view;
    //     readonly    attribute long    detail;
    // };
    Event.call(this, ev);
    
    
    /**
     * @name ScrollEvent#scrollTop
     * @type Number
     */
    this.scrollTop = ev.target.scrollTop;

    /**
     * @name ScrollEvent#scrollLeft
     * @type Number
     */
    this.scrollLeft =  ev.target.scrollLeft;

    /**
     * @name ScrollEvent#scrollHeight
     * @type Number
     */
    this.scrollHeight = ev.target.scrollHeight;
    /**
     * @name ScrollEvent#detail
     * @type Number
     */
    this.detail = ev.detail;
}

ScrollEvent.prototype = Object.create(Event.prototype);
ScrollEvent.prototype.constructor = ScrollEvent;

/**
 * Return the name of the event type
 *
 * @method
 *
 * @return {String} Name of the event type
 */
ScrollEvent.prototype.toString = function toString () {
    return 'ScrollEvent';
};

module.exports = ScrollEvent;
